### DISCUSSION TITLE
Longhorn can't attach to pod nor node.

### DISCUSSION THREAD

**Discussion by OrvilleQ** [longhorn/longhorn#5367] [Q&A]
Hello there.

I just installed longhorn 1.4.0 on k3s v1.26.1+k3s1 with the default helm settings. The host os is Fedora CoreOS,  SELinux for k3s is configured and no AVC or SELINUX_ERR are reported. ALL pod inside longhorn-system are healthy but I can't create pod with pvc bound to it. The pod will stuck at ContainerCreating or Terminating phase with
``` 
Warning  FailedAttachVolume  48s   attachdetach-controller  AttachVolume.Attach failed for volume "pvc-12e35c5f-bf7f-4357-a374-dede3c0ed250" : rpc error: code = DeadlineExceeded desc = volume pvc-12e35c5f-bf7f-4357-a374-dede3c0ed250 failed to attach to node sea01.svr.anislet.dev
```

In the ui, all three replicas are showing as running during the attaching phase but never been attached the to pod. I can't manually attach PVC to node either with 1\2\3 replicas.

Here is the support bundle: 

[supportbundle_9fee904e-f2dc-4147-a86f-1fa12e5dba27_2023-02-08T05-03-40Z.zip](https://github.com/longhorn/longhorn/files/10681942/supportbundle_9fee904e-f2dc-4147-a86f-1fa12e5dba27_2023-02-08T05-03-40Z.zip)

Thank you for your help.

**Comment by innobead** [comment]
>  2023-02-08T05:03:30.925821719Z time="2023-02-08T05:03:30Z" level=warning msg="pvc-12e35c5f-bf7f-4357-a374-dede3c0ed250-e-80f2c5dc: time=\"2023-02-08T04:52:29Z\" level=error msg=\"Failed to startup frontend\" error=\"failed to execute: nsenter [--mount=/host/proc/1/ns/mnt --net=/host/proc/1/ns/net iscsiadm -m node -T iqn.2019-10.io.longhorn:pvc-12e35c5f-bf7f-4357-a374-dede3c0ed250 -o update -n node.session.err_timeo.abort_timeout -v 15], output , stderr iscsiadm: No records found\\n: exit status 21\""

It seems the volume frontend target not found :thinking: , but from the code, it should be already discovered before calling `UpdateIscsiDeviceAbortTimeout`. 

[/iscsidev/iscsi.go#L146-L166](https://github.com/longhorn/go-iscsi-helper/blob/88c602ba3096f71f9ed426b252de7d38770a05b2/iscsidev/iscsi.go#L146-L166)
```go
	for i := 0; i < RetryCounts; i++ {
		err := iscsi.DiscoverTarget(localIP, dev.Target, ne)
		if iscsi.IsTargetDiscovered(localIP, dev.Target, ne) {
			break
		}

		logrus.WithError(err).Warnf("Failed to discover")
		// This is a trick to recover from the case. Remove the
		// empty entries in /etc/iscsi/nodes/<target_name>. If one of the entry
		// is empty it will triggered the issue.
		if err := iscsi.CleanupScsiNodes(dev.Target, ne); err != nil {
			logrus.WithError(err).Warnf("Failed to clean up nodes for %v", dev.Target)
		} else {
			logrus.Warnf("Nodes cleaned up for %v", dev.Target)
		}

		time.Sleep(RetryIntervalSCSI)
	}
	if err := iscsi.UpdateIscsiDeviceAbortTimeout(dev.Target, dev.IscsiAbortTimeout, ne); err != nil {
		return err
	}
```

ref: 21 ISCSI_ERR_NO_OBJS_FOUND - no records/targets/sessions/portals found to execute operation on.

@derekbit @shuo-wu do we have tgtd log collected? any situation will cause this?

**Reply by shuo-wu** [reply]
Longhorn won't collect iscsid log now. But the error log of tgt part will be recorded in the instance managers as well.

**Reply by derekbit** [reply]
> @derekbit @shuo-wu do we have tgtd log collected? any situation will cause this?

No.


The logic is a bit of weird but not harmful. The error messages show the target is not discovered but does not return error here. The error is returned in `UpdateIscsiDeviceAbortTimeout` in the end.

**Comment by derekbit** [comment]
@OrvilleQ 
Can you enter the node `sea01.svr.anislet.dev` and list the sessions by `iscsiadm -m session`?
If `iqn.2019-10.io.longhorn:pvc-12e35c5f-bf7f-4357-a374-dede3c0ed250` exists, see if the command `iscsiadm -m node -T iqn.2019-10.io.longhorn:pvc-12e35c5f-bf7f-4357-a374-dede3c0ed250 -o update -n node.session.err_timeo.abort_timeout -v 15` is allowable

**Reply by OrvilleQ** [reply]
`iscsiadm -m session` shows no active session.

**Reply by OrvilleQ** [reply]
```
[root@sea01 ~]# iscsiadm -m session
iscsiadm: No active sessions.
[root@sea01 ~]# systemctl status iscsid.service 
● iscsid.service - Open-iSCSI
     Loaded: loaded (/usr/lib/systemd/system/iscsid.service; enabled; preset: disabled)
     Active: active (running) since Wed 2023-02-08 03:08:27 UTC; 6h ago
TriggeredBy: ● iscsid.socket
       Docs: man:iscsid(8)
             man:iscsiuio(8)
             man:iscsiadm(8)
   Main PID: 933 (iscsid)
     Status: "Ready to process requests"
      Tasks: 1 (limit: 9433)
     Memory: 3.3M
        CPU: 1.756s
     CGroup: /system.slice/iscsid.service
             └─933 /usr/sbin/iscsid -f

Feb 08 03:08:27 sea01.svr.anislet.dev systemd[1]: Starting iscsid.service - Open-iSCSI...
Feb 08 03:08:27 sea01.svr.anislet.dev systemd[1]: Started iscsid.service - Open-iSCSI.
[root@sea01 ~]# lsmod | grep iscsi
iscsi_tcp              32768  0
libiscsi_tcp           36864  1 iscsi_tcp
libiscsi               86016  2 libiscsi_tcp,iscsi_tcp
scsi_transport_iscsi   155648  4 libiscsi_tcp,iscsi_tcp,libiscsi
[root@sea01 ~]# ls /var/lib/iscsi/nodes/
iqn.2019-10.io.longhorn:321  iqn.2019-10.io.longhorn:pvc-12e35c5f-bf7f-4357-a374-dede3c0ed250
```

**Reply by OrvilleQ** [reply]
Wait, my bad. There is some SELinux AVC messages:
`time->Wed Feb  8 09:14:19 2023
type=AVC msg=audit(1675847659.049:11085): avc:  denied  { dac_override } for  pid=409824 comm="iscsiadm" capability=1  scontext=system_u:system_r:iscsid_t:s0 tcontext=system_u:system_r:iscsid_t:s0 tclass=capability permissive=0
`

So this might be a SELinux related issue?

**Reply by OrvilleQ** [reply]
After `setenforce 0`, The volume do successfully attached.

**Reply by derekbit** [reply]
> So this might be a SELinux related issue?

I think so. Probably related to this issue https://github.com/longhorn/longhorn/issues/4988.
You can try the workaround https://github.com/longhorn/longhorn/issues/4988#issuecomment-1345441977.

**Reply by OrvilleQ** [reply]
> iscsi-initiator-utils 6.2.1.4 is based on open-iscsi 2.1.4. This issue appears to be caused by the permissions of the directories. It has been fixed in https://github.com/open-iscsi/open-iscsi/pull/244 and is included in open-iscsi 2.1.4.

It seems like if I'm using open-iscsi larger than 2.1.4 things should be fine. Though the recent `iscsi-initiatoe-utils` is based on open-iscsi 2.1.8, same error still occur. That's really wired.

Thanks for your help.

**Reply by innobead** [reply]
cc @ChanYiLin to help with creating a KB. (I feel we need to explain a bit about security policy impact when using Longhorn as well at https://longhorn.io/docs/1.4.0/deploy/install/#installation-requirements)

**Reply by OrvilleQ** [reply]
Had a hard time trying to find out which directory or file is this `DAC_OVERRIDE` error targeted on, but still got nothing. Do you guys have any idea which file might behind this issue since give iscsiadm `CAP_DAC_OVERRIDE` is just a lazy way which makes iscsid can access everything, which I want to avoid if possible.

Thank you.

**Reply by derekbit** [reply]
IIRC, Fedora CoreOS, centos stream and fedora enable SELinux by default and are popular OSes. Although we have a KB for OKD (https://longhorn.io/kb/troubleshooting-volumes-stuck-in-attach-detach-loop-when-using-longhorn-on-okd/), it is worth investigating and figuring out a solution for SELinux enabled env.

cc @innobead @ChanYiLin

**Reply by derekbit** [reply]
Another SELinux enabled case on CentOS stream 9
https://cloud-native.slack.com/archives/CNVPEL9U3/p1679321973634209

**Reply by innobead** [reply]
@khushboo-rancher I know we have some SELinux test cases. What test cases are covered now?

**Reply by innobead** [reply]
Just created an issue to follow up. https://github.com/longhorn/longhorn/issues/5627

**Reply by davidgaster** [reply]
Hello! We are deploying 3 node RKE2 clusters. It seems to work fine on a single node, however when setting up HA RKE2 with 3 nodes, we are running into this exact issue. Disabling SELinux resolves the problem and PVs are then able to attach to Statefulsets. However, our customers will need SELinux enabled and a working solution. Any update on this issue? 

Same error in Instance Manager logs:
```
stderr iscsiadm: No records found\n: exit status 21
``` 
```
$ iscsiadm -m session
iscsiadm: No active sessions.
```

OS, RKE2, Longhorn
```
Red Hat Enterprise Linux 8.7 (Ootpa)
v1.24.12+rke2r1", GitCommit:"ef70d260f3d036fc22b30538576bbf6b36329995", GitTreeState:"clean"
Longhorn 1.4
```

**Reply by khushboo-rancher** [reply]
@innobead We have CI to run e2e test cases on SELinux RHEL. However, we have it ran on 8.6 version for the v1.4.0 release. 
I'm running the test with 1.4.x to double check it and will work on bumping the version to 8.7 and run the test on it later.

**Reply by davidgaster** [reply]
Thanks! Curious do the e2e test cases include multi-node? We don't run into this issue on single node Longhorn. Also would be nice to surface the issue more easily

**Reply by khushboo-rancher** [reply]
Our e2e test cases run on 3 worker nodes cluster. The result of running tests on RHEL 8.6.0 SELinux enabled looks normal.
